include/zephyr/kernel/obj_core.h

```diff
@@ -0,0 +1,283 @@
+	k_obj_core_init(_objp, &_obj_type)
+
+#define K_OBJ_CORE_LINK(objp) k_obj_core_link(objp)
+#else
+#define K_OBJ_CORE_INIT(objp, type)   do { } while(0)
+#define K_OBJ_CORE_LINK(objp)         do { } while(0)
+#endif
+/**
+ * Tools may use this list as an entry point to identify all registered
+ * object types and the object cores linked to them.
+ */
+extern sys_dlist_t obj_type_list;
+
+struct k_obj_type {
+	sys_dnode_t    node;   /* Node within list of object types */
+	sys_dlist_t    list;   /* List of objects of this object type */
+	uint32_t       id;     /* Unique type ID */
+	size_t         obj_core_offset;  /* Offset to obj_core field */
+};
+
+struct k_obj_core {
+	sys_dnode_t        node;   /* Object node within object type's list */
+	struct k_obj_type *type;   /* Object type to which object belongs */
+};
+
+/**
+ * @brief Initialize a specific object type
+ *
+ * Initializes a specific object type and links it into the object core
+ * framework.
+ *
+ * @param type Pointer to the object type to initialize
+ * @param id A means to identify the object type
+ * @param off Offset of object core within the structure
+ */
+extern void z_obj_type_init(struct k_obj_type *type, uint32_t id, size_t off);
+
+/**
+ * @brief Find a specific object type by ID
+ *
+ * Given an object type ID, this function searches for the object type that
+ * is associated with the specified type ID @a type_id.
+ *
+ * @param type_id  Type ID associated with object type
+ *
+ * @retval NULL if object type not found
+ * @retval Pointer to object type if found
+ */
+extern struct k_obj_type *k_obj_type_find(uint32_t type_id);
+
+/**
+ * @brief Walk the object type's list of object cores
+ *
+ * This function walks the object type's list of object cores and invokes the
+ * specified callback function on each element while locked. Although this will
+ * ensure that that list is not modified, one can expect a significant penalty
+ * in terms of performance and latency.
+ *
+ * The callback function shall either return non-zero to stop further walking,
+ * or it shall return 0 to continue walking.
+ *
+ * @param type  Pointer to the object type
+ * @param func  Callback to invoke on each object core of the object type
+ * @param data  Custom data passed to the callback
+ *
+ * @retval non-zero if walk is terminated by the callback; otherwise 0
+ */
+extern int k_obj_type_walk_locked(struct k_obj_type *type,
+				  int (*func)(struct k_obj_core *, void *),
+				  void *data);
+
+/**
+ * @brief Walk the object type's list of object cores
+ *
+ * This function walks the object type's list of object cores and invokes the
+ * specified callback function on each element while unlocked. This routine
+ * does not prevent the addition or removal of object cores while walking the
+ * list. Although this is expected to offer better responsiveness, it may
+ * result in unexpected behavior. Note that locks are enabled while invoking
+ * the callback.
```

Comment:
Same comment: what exactly does "locks are enabled" mean?

Reply:
I will need to clarify.